packages/nimble-components/src/table/specs/table-column-custom-sort-field-hld.md

```diff
@@ -0,0 +1,31 @@
+# Custom table column sort order HLD
+
+## Problem Statement
+
+In some cases, a client may have the need to specify a custom sort order for a column. For example, a table displaying SystemLink tags may have a field in their records for the stringified tag but may not want to sort that column that stringified value. Instead, they want to provide their own sort order for tags that takes each tag's data type and value into account.
+
+## Links to Relevant Work Items and Reference Material
+
+[Azdo Feature 2665131 - Custom column sorting](https://dev.azure.com/ni/DevCentral/_workitems/edit/2665131)
+
+## Implementation / Design
+
+Columns that need support for a custom column sort order will extend their API to allow a client to optionally specify an additional field within the data that specifies the sort order of that column. This field is expected to be a numeric value that is the sorted data index of the record. When the table is sorted by this column, the table will use the `basic` sort operation and sort the records by the additionally specified field rather than the default `operandDataRecordFieldName` associated with the column.
+
+TanStack uses the same field internally for both sorting and grouping. Grouping by this custom sort field could likely lead to unexpected results since there isn't a guarantee that two values that render identically have the same sort sort index. Therefore, grouping and specifying a custom sort order will not be allowed on a column at the same time. If both are specified at the same time, the column will be considered invalid and the `invalidCustomSortWithGrouping` validation flag will be set on the column.
```

Comment:
Another possibility might be to respond to setting the group index on the column by switching which column TanStack uses for grouping/sorting. I.e. when grouping, use the default `operandDataRecordFieldName`, otherwise use the `sort-by-field-name`. We still couldn't support both sorting _and_ grouping at the same time.

Reply:
Yeah, technically we could do that, but I'm not sure how much that gets us.

Actually, if we really wanted to support grouping & sorting together on these columns we probably could do that. We'd just need to create two different TanStack columns for a given nimble column and then manage the state of each one. I'll add that as an alternative so that we can decide if we can weigh the additional complexity vs the additional functionality.

Reply:
Let's see what others think, but if it's pretty cheap to support grouping and sorting (but not at the same time), that seems worthwhile. Supporting both at the same time sounds like it might add more complexity than it's worth.

Reply:
My 2 cents: 
If we really wanted both sorting and grouping, one option would be to put the burden on the client to provide equal `sort-by-field-name` values for records that should be grouped together. Sorting would fall back to whatever tiebreaker we decide (see my other thread) and grouping would behave as it does today. A downside of that would be increased complexity on client code for generating the sort by field values.

If we don't like that increased complexity (e.g. if we think clients will get it wrong) then I would lean towards not supporting grouping at all. It's not worth adding complexity if we don't have a requirement to allow sorting and grouping together. And for the types of mixed and not-well-normalized data in the column types we know about, grouping by the display value won't be particularly useful (e.g. separate groups for tags "1", "1.0", and "1%").

Reply:
Your suggestion of grouping on `sort-by-field-name` doesn't seem unreasonable from a behavior perspective. However, that still isn't free from the nimble side. Specifically, the group row's value comes from TanStack, but in your suggestion that would be the sort index (e.g. `0`, `1`, `2`, etc), which is not what we would want to have render in the group row. Therefore, we'd still have to add code that can figure out what the "actual" field value is, rather than the sort index.